### ISSUE TITLE
Rule Change: Avoid highlighting the entire file for max-classes-per-file violations

### ISSUE THREAD

**Issue by Pixel998** [eslint/eslint#20968]
### What rule do you want to change?

max-classes-per-file

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/*eslint max-classes-per-file: "error"*/

class Foo {}

class Bar {}
```

### What does the rule currently do for this code?

Currently, the rule reports on the `Program` node, so editors will highlight the entire file. This is noisy because the violation is caused by the presence of an additional class, not by every line in the file.

### What will the rule do after it's changed?

Ideally, the rule would report on a smaller, more precise location, such as the class head for the class that exceeds the configured limit. For example, in the code above, the violation could be reported on `class B`.

Another alternative would be to keep reporting from `Program:exit`, but use the same effective range that `Program` had before the v10 range change:

```js
loc: {
    start: node.body[0].loc.start,
    end: node.body.at(-1).loc.end,
}
```

This would not be as precise as reporting on the class that exceeds the limit, because it would still cover the program body. However, it would avoid including leading/trailing comments in the reported range and would restore behavior closer to v9. For example:

```js
/* leading file comment */
/* eslint-disable max-classes-per-file */

class A {}
class B {}
```

In v9, this disable comment suppresses the violation. In v10, it does not.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by mdjermanovic** [comment]
> Another alternative would be to keep reporting from `Program:exit`, but use the same effective range that `Program` had before the v10 range change:
>
> ```js
> loc: {
>     start: node.body[0].loc.start,
>     end: node.body.at(-1).loc.end,
> }
> ```
> This would not be as precise as reporting on the class that exceeds the limit, because it would still cover the program body. However, it would avoid including leading/trailing comments in the reported range and would restore behavior closer to v9. For example:
> 
> ```js
> /* leading file comment */
> /* eslint-disable max-classes-per-file */
> 
> class A {}
> class B {}
> ```
> In v9, this disable comment suppresses the violation. In v10, it does not.

This makes sense to me, as a bug fix that would be similar to https://github.com/eslint/eslint/issues/20451.

> Ideally, the rule would report on a smaller, more precise location, such as the class head for the class that exceeds the configured limit.

I think this is worth considering, but as described in https://github.com/eslint/eslint/pull/20943#issuecomment-4620538335 it would be a breaking change.

@eslint/eslint-team thoughts?

**Comment by DMartens** [comment]
I agree with the change but would defer it to the next major version.
This breaking change affects more files than I would thought ([github search](https://github.com/search?q=eslint-disable+max-classes-per-file&type=code)).

**Comment by lumirlumir** [comment]
According to the consensus, I’m moving it to the v11 triage board and marking it as blocked.
